cow_instrument/DetectorInfo.h

```diff
@@ -181,13 +229,72 @@ DetectorInfo<InstTree>::DetectorInfo(std::shared_ptr<InstTree> &&instrumentTree)
       m_positions(std::make_shared<std::vector<Eigen::Vector3d>>(m_nDetectors)),
       m_rotations(
           std::make_shared<std::vector<Eigen::Quaterniond>>(m_nDetectors)),
+      m_linearIndexMap(makeDefaultIndexes(instrumentTree)),
+      m_durations(std::make_shared<const ScanTimes>(1, scanTime)),
       m_pathComponentInfo(
           std::forward<std::shared_ptr<InstTree>>(instrumentTree)) {
 
   init();
 }
 
 template <typename InstTree>
+template <typename InstSptrType, typename TimeIndexesType,
+          typename ScanTimesType, typename PositionsType,
+          typename RotationsType>
+DetectorInfo<InstTree>::DetectorInfo(InstSptrType &&instrumentTree,
+                                     TimeIndexesType &&timeIndexes,
+                                     ScanTimesType &&scanTimes,
+                                     PositionsType &&positions,
+                                     RotationsType &&rotations)
+    : m_l2Paths(SourceSampleDetectorPathFactory<InstTree>{}.createL2(
+          *instrumentTree)),
+      m_l1Paths(SourceSampleDetectorPathFactory<InstTree>{}.createL1(
+          *instrumentTree)),
+      m_nDetectors(instrumentTree->nDetectors()),
+      m_l1(std::make_shared<L1s>(m_nDetectors)),
+      m_l2(std::make_shared<L2s>(m_nDetectors * scanTimes.size())),
+      m_isMasked(std::make_shared<MaskFlags>(m_nDetectors, Bool(false))),
+      m_isMonitor(std::make_shared<MonitorFlags>(m_nDetectors, Bool(false))),
+      m_detectorComponentIndexes(std::make_shared<const std::vector<size_t>>(
+          instrumentTree->detectorComponentIndexes())),
+      m_positions(std::make_shared<std::vector<Eigen::Vector3d>>(
+          std::forward<PositionsType>(positions))),
+      m_rotations(std::make_shared<std::vector<Eigen::Quaterniond>>(
+          std::forward<RotationsType>(rotations))),
+      m_linearIndexMap(std::make_shared<std::vector<std::vector<size_t>>>(
+          std::forward<TimeIndexesType>(timeIndexes))),
+      m_durations(
+          std::make_shared<ScanTimes>(std::forward<ScanTimesType>(scanTimes))),
+      m_pathComponentInfo(
+          std::forward<std::shared_ptr<InstTree>>(instrumentTree)),
+      m_isScanning(true) {
+  size_t indexEntryCount = 0;
+  for (auto &positionTimes : *m_linearIndexMap) {
+    // Count up the time indexes for each component
+    indexEntryCount += positionTimes.size();
+  }
+  if (indexEntryCount != m_positions->size()) {
```

Comment:
I think this would still work. But I think you've discovered two bugs.

1. DetectorInfo::makeSpectra() should be making a spectra for every `position` not based on the `detectorSize` (as it currently does.
1. DetectorInfo::initL2() should be making the same number of `l2s` as we have positions in the scan. I am currently creating `l2s` that aren't required.

With those two in place you should be able to do something like:

```cpp
DetectorInfo detInfo(instrument, positions, rotations);
SpectrumInfo spectrumInfo(detInfo);
```